src/cassandra/future.rs

```diff
@@ -148,29 +149,31 @@ impl<T: Completable> futures::Future for CassFuture<T> {
                         // already, complete now without scheduling a callback.
                         if cass_future_ready(self.inner) == cass_true {
                             // Future is ready; wrap success in `Ok(Ready)` or report failure as `Err`.
-                            get_completion(self.inner).map(futures::Async::Ready)
+                            Poll::Ready(get_completion(self.inner))
                         } else {
                             // Future is not ready; park this task and arrange to be called back when
                             // it is.
                             *state = FutureState::Awaiting {
-                                task: futures::task::current(),
+                                waker: cx.waker().clone(),
                                 keep_alive: self.state.clone(),
                             };
                             install_callback = true;
-                            Ok(futures::Async::NotReady)
+                            Poll::Pending
                         }
                     }
-                    FutureState::Awaiting { ref mut task, .. } => {
+                    FutureState::Awaiting { ref mut waker, .. } => {
                         // Callback already scheduled; don't set it again (C doesn't support it anyway),
                         // but be sure to swizzle the new task into place. No need to check for
                         // readiness here; we have to wait for the callback anyway so we might as well
                         // do all the work in one place.
-                        *task = futures::task::current();
-                        Ok(futures::Async::NotReady)
+                        if !waker.will_wake(cx.waker()) {
```

Comment:
This is a change to the existing code. Is it necessary, and why is it correct/safe? The comment says "be sure to swizzle the new task into place", but it looks like we now do *not* necessarily swizzle the new task into place.

(Not saying the old code was necessarily correct either, but it naively looks like the new code differs - though maybe not essentially. Really I'm just looking for an explanation, since I'm not yet familiar with futures 0.3.) Thanks.

Reply:
Right, so, each time the future is polled, it may or may not get a new waker. This is purely an optimization though. What we are saying is "is the waker that is already in place the same as the one that we were just called with? if it is, then we don't need to do anything, otherwise, if it isn't, we need to swap the waker out." 

So, we're saying "if the waker we gave to C++ land would wake me up, then I have nothing to do, otherwise I have to give it a new waker in order to wake me up." 

This is purely an optimization, and is mentioned in the std docs here: https://doc.rust-lang.org/beta/std/task/struct.Waker.html#method.will_wake

Reply:
Cool, that makes sense. I thought that was probably it, but wanted to confirm.